README.md
# @nyxb/colors

The tiniest and the fastest library for terminal output formatting with ANSI && RGB colors.

```javascript
import nc from "@nyxb/colors"

console.log(
  nc.green(`How are ${c.italic(`you`)} doing?`)
)
```

- **No dependencies.**
- **14 times** smaller and **2 times** faster than chalk.
- Used by popular tools like PostCSS, SVGO, Stylelint, and Browserslist.
- Node.js v6+ & browsers support. Support for both CJS and ESM projects.
- TypeScript type declarations included.
- [`NO_COLOR`](https://no-color.org/) friendly.

## Motivation

With `@nyxb/colors` we are trying to draw attention to the `node_modules` size
problem and promote performance-first culture.

## Prior Art

Credits go to the following projects:

- [Nanocolors](https://github.com/ai/nanocolors) by [@ai](https://github.com/ai)
- [Colorette](https://github.com/jorgebucaran/colorette) by [@jorgebucaran](https://github.com/jorgebucaran)
- [Kleur](https://github.com/lukeed/kleur) by [@lukeed](https://github.com/lukeed)
- [Colors.js](https://github.com/Marak/colors.js) by [@Marak](https://github.com/Marak)
- [Chalk](https://github.com/chalk/chalk) by [@sindresorhus](https://github.com/sindresorhus)

## Benchmarks

The space in node_modules including sub-dependencies:

```diff
$ node ./benchmarks/size.js
Data from packagephobia.com
  chalk        43 kB
  cli-color   796 kB
  ansi-colors  27 kB
  kleur        20 kB
  colorette    17 kB
  nanocolors   15 kB
+ colors       39 kB
```

Library loading time:

```diff
$ node ./benchmarks/loading.js
  chalk          3.024 ms
  cli-color     18.894 ms
  ansi-colors    0.821 ms
  kleur          1.171 ms
  kleur/colors   0.581 ms
  colorette      0.529 ms
  nanocolors     0.319 ms
+ colors         0.157 ms
```

Benchmark for simple use case:

```diff
$ node ./benchmarks/simple.js
  chalk         48,283,199 ops/sec
  cli-color      2,591,647 ops/sec
  ansi-colors   13,762,906 ops/sec
  kleur         52,427,142 ops/sec
  kleur/colors  66,904,098 ops/sec
  colorette     66,570,482 ops/sec
  nanocolors    87,642,042 ops/sec
+ colors        86,856,783 ops/sec
```

Benchmark for complex use cases:

```diff
$ node ./benchmarks/complex.js
  chalk          2,113,896 ops/sec
  cli-color        292,431 ops/sec
  ansi-colors    1,117,678 ops/sec
  kleur          2,397,329 ops/sec
  kleur/colors   2,685,299 ops/sec
  colorette      2,907,371 ops/sec
  nanocolors     3,119,996 ops/sec
+ colors         5,292,435 ops/sec
```

## Usage

@nyxb/colors provides an object which includes a variety of text coloring and formatting functions

```javascript
import nc from "@nyxb/colors"
```

The object includes following coloring functions: `black`, `red`, `green`, `yellow`, `blue`, `magenta`, `cyan`, `white`, `gray` and special colors from nyxb brand `nyxbgreen` `nyxbblue` `nyxbfox` `nyxbpurple` `nyxbyellow` `nyxbred` `nyxbcyan`.

```javascript
console.log(`I see a ${nc.red("red door")} and I want it painted ${nc.black("black")}`)
```

The object also includes following background color modifier functions: `bgBlack`, `bgRed`, `bgGreen`, `bgYellow`, `bgBlue`, `bgMagenta`, `bgCyan`, `bgWhite`.

```javascript
console.log(
  nc.bgBlack(
    nc.white(`Tom appeared on the sidewalk with a bucket of whitewash and a long-handled brush.`)
  )
)
```

Besides colors, the object includes following formatting functions: `dim`, `bold`, `hidden`, `italic`, `underline`, `strikethrough`, `reset`, `inverse`.

```javascript
for (let task of tasks) {
  console.log(`${nc.bold(task.name)} ${nc.dim(task.durationMs + "ms")}`)
}
```

The library provides additional utilities to ensure the best results for the task:

- `isColorSupported` — boolean, explicitly tells whether or not the colors or formatting appear on the screen

  ```javascript
  import nc from "@nyxb/colors"

  if (nc.isColorSupported) {
    console.log("Yay! This script can use colors and formatters")
  }
  ```

- `createColors(enabled)` — a function that returns a new API object with manually defined color support configuration

  ```javascript
  import nc from "@nyxb/colors"

  let { red, bgWhite } = nc.createColors(options.enableColors)
  ```

## Replacing `chalk`

1. Replace package name in import:

   ```diff
   - import chalk from 'chalk'
   + import nyxc from '@nyxb/colors'
   ```

2. Replace variable:

   ```diff
   - chalk.red(text)
   + nyxc.red(text)
   ```

3. Replace chains to nested calls:

   ```diff
   - chalk.red.bold(text)
   + nyxc.red(nyxc.bold(text))
   ```

4. You can use [`colorize-template`](https://github.com/usmanyunusov/colorize-template)
   to replace chalk’s tagged template literal.

   ```diff
   + import { createColorize } from 'colorize-template'

   + let colorize = createColorize(nyxc)
   - chalk.red.bold`full {yellow ${"text"}}`
   + colorize`{red.bold full {yellow ${"text"}}}`
   ```


CHANGELOG.md
# Changelog


## v0.0.3

[compare changes](https://github.com/nyxb/colors/compare/v0.0.2...v0.0.3)

## v0.0.2



tests/test.js
const assert = require('node:assert')
const consolji = require('consolji')
const nc = require('../colors.js')

const FMT = {
   reset: ['\x1B[0m', '\x1B[0m'],
   bold: ['\x1B[1m', '\x1B[22m'],
   dim: ['\x1B[2m', '\x1B[22m'],
   italic: ['\x1B[3m', '\x1B[23m'],
   underline: ['\x1B[4m', '\x1B[24m'],
   inverse: ['\x1B[7m', '\x1B[27m'],
   hidden: ['\x1B[8m', '\x1B[28m'],
   strikethrough: ['\x1B[9m', '\x1B[29m'],
   black: ['\x1B[30m', '\x1B[39m'],
   red: ['\x1B[31m', '\x1B[39m'],
   green: ['\x1B[32m', '\x1B[39m'],
   yellow: ['\x1B[33m', '\x1B[39m'],
   blue: ['\x1B[34m', '\x1B[39m'],
   magenta: ['\x1B[35m', '\x1B[39m'],
   cyan: ['\x1B[36m', '\x1B[39m'],
   white: ['\x1B[37m', '\x1B[39m'],
   gray: ['\x1B[90m', '\x1B[39m'],
   nyxbpurple: ['\x1B[38;2;153;69;255m', '\x1B[39m'],
   nyxbgreen: ['\x1B[38;2;20;241;149m', '\x1B[39m'],
   nyxbcyan: ['\x1B[38;2;0;255;255m', '\x1B[39m'],
   nyxbblue: ['\x1B[38;2;88;101;242m', '\x1B[39m'],
   nyxbyellow: ['\x1B[38;2;254;231;92m', '\x1B[39m'],
   nyxbfox: ['\x1B[38;2;235;69;158m', '\x1B[39m'],
   nyxbred: ['\x1B[38;2;237;66;69m', '\x1B[39m'],
   bgBlack: ['\x1B[40m', '\x1B[49m'],
   bgRed: ['\x1B[41m', '\x1B[49m'],
   bgGreen: ['\x1B[42m', '\x1B[49m'],
   bgYellow: ['\x1B[43m', '\x1B[49m'],
   bgBlue: ['\x1B[44m', '\x1B[49m'],
   bgMagenta: ['\x1B[45m', '\x1B[49m'],
   bgCyan: ['\x1B[46m', '\x1B[49m'],
   bgWhite: ['\x1B[47m', '\x1B[49m'],
}

test('color matching', () => {
   for (const format in FMT) {
      assert.equal(nc[format]('string'), `${FMT[format][0]}string${FMT[format][1]}`)
      consolji.log(nc[format](`testing: ${format}`))
   }
})

test('format/color nesting', () => {
   assert.equal(
      nc.bold(`BOLD ${nc.red(`RED ${nc.dim('DIM')} RED`)} BOLD`),
      `${FMT.bold[0]
         }BOLD ${
         FMT.red[0]
         }RED ${
         FMT.dim[0]
         }DIM${
         FMT.dim[1]
         }${FMT.bold[0]
         } RED${
         FMT.red[1]
         } BOLD${
         FMT.bold[1]}`,
   )
})

test('proper wrapping', () => {
   assert.equal(
      nc.red(nc.bold('==TEST==')),
      `${FMT.red[0] + FMT.bold[0]}==TEST==${FMT.bold[1]}${FMT.red[1]}`,
   )
})

test('complex case of wrapping', () => {
   assert.equal(
      nc.bold(nc.yellow(nc.bgRed(nc.italic('==TEST==')))),
      `${FMT.bold[0]
         + FMT.yellow[0]
         + FMT.bgRed[0]
         + FMT.italic[0]
         }==TEST==${
         FMT.italic[1]
         }${FMT.bgRed[1]
         }${FMT.yellow[1]
         }${FMT.bold[1]}`,
   )

   assert.equal(
      nc.cyan(nc.bold(nc.underline('==TEST=='))),
      `${FMT.cyan[0]
         + FMT.bold[0]
         + FMT.underline[0]
         }==TEST==${
         FMT.underline[1]
         }${FMT.bold[1]
         }${FMT.cyan[1]}`,
   )
})

test('close sequence replacement', () => {
   assert.equal(
      nc.red(`foo ${nc.yellow('bar')} baz`),
      `${FMT.red[0]}foo ${FMT.yellow[0]}bar${FMT.red[0]} baz${FMT.red[1]}`,
   )

   assert.equal(
      nc.bold(`foo ${nc.red(nc.dim('bar'))} baz`),
      `${FMT.bold[0]
         }foo ${
         FMT.red[0]
         }${FMT.dim[0]
         }bar${
         FMT.dim[1]
         }${FMT.bold[0]
         }${FMT.red[1]
         } baz${
         FMT.bold[1]}`,
   )

   assert.equal(
      nc.yellow(`foo ${nc.red(nc.bold('red'))} bar ${nc.cyan('cyan')} baz`),
      `${FMT.yellow[0]
         }foo ${
         FMT.red[0]
         }${FMT.bold[0]
         }red${
         FMT.bold[1]
         }${FMT.yellow[0]
         } bar ${
         FMT.cyan[0]
         }cyan${
         FMT.yellow[0]
         } baz${
         FMT.yellow[1]}`,
   )
})

test('non-string input', () => {
   assert.equal(nc.red(), `${FMT.red[0]}undefined${FMT.red[1]}`)
   assert.equal(nc.red(undefined), `${FMT.red[0]}undefined${FMT.red[1]}`)
   assert.equal(nc.red(0), `${FMT.red[0]}0${FMT.red[1]}`)
   assert.equal(nc.red(NaN), `${FMT.red[0]}NaN${FMT.red[1]}`)
   assert.equal(nc.red(null), `${FMT.red[0]}null${FMT.red[1]}`)
   assert.equal(nc.red(true), `${FMT.red[0]}true${FMT.red[1]}`)
   assert.equal(nc.red(false), `${FMT.red[0]}false${FMT.red[1]}`)
   assert.equal(nc.red(Infinity), `${FMT.red[0]}Infinity${FMT.red[1]}`)
})

function test(name, fn) {
   try {
      fn()
      consolji.log(nc.green(`✓ ${name}`))
   }
   catch (error) {
      consolji.log(nc.red(`✗ ${name}`))
      throw error
   }
}


tests/environments.js
const vm = require('node:vm')
const fs = require('node:fs')
const assert = require('node:assert')
const consolji = require('consolji')
const nc = require('../colors.js')

const source = fs.readFileSync(`${__dirname}/../colors.js`, 'utf-8')

test('ci server', () => {
   const nc = initModuleEnv({ env: { TERM: 'dumb', CI: '1' } })
   assert.equal(nc.isColorSupported, true)
   assert.equal(nc.red('text'), nc.createColors(true).red('text'))
})

test('arg --color', () => {
   const nc = initModuleEnv({ env: { TERM: 'dumb' }, argv: ['--color'] })
   assert.equal(nc.isColorSupported, true)
   assert.equal(nc.red('text'), nc.createColors(true).red('text'))
})

test('env NO_COLOR', () => {
   const nc = initModuleEnv({ env: { FORCE_COLOR: '1', NO_COLOR: '1' } })
   assert.equal(nc.isColorSupported, false)
   assert.equal(nc.red('text'), nc.createColors(false).red('text'))
})

test('env FORCE_COLOR', () => {
   const nc = initModuleEnv({ env: { TERM: 'dumb', FORCE_COLOR: '1' } })
   assert.equal(nc.isColorSupported, true)
   assert.equal(nc.red('text'), nc.createColors(true).red('text'))
})

test('arg --no-color', () => {
   const nc = initModuleEnv({ env: { FORCE_COLOR: '1' }, argv: ['--no-color'] })
   assert.equal(nc.isColorSupported, false)
   assert.equal(nc.red('text'), nc.createColors(false).red('text'))
})

test('no term', () => {
   const nc = initModuleEnv({ env: { TERM: 'dumb' } })
   assert.equal(nc.isColorSupported, false)
   assert.equal(nc.red('text'), nc.createColors(false).red('text'))
})

test('windows', () => {
   const nc = initModuleEnv({ env: { TERM: 'dumb' }, platform: 'win32' })
   assert.equal(nc.isColorSupported, true)
   assert.equal(nc.red('text'), nc.createColors(true).red('text'))
})

function test(name, fn) {
   try {
      fn()
      consolji.log(nc.green(`✓ ${name}`))
   }
   catch (error) {
      consolji.log(nc.red(`✗ ${name}`))
      throw error
   }
}

function initModuleEnv({ env, argv = [], platform = 'darwin' }) {
   const process = { env, argv, platform }
   const context = vm.createContext({ require, process, module: { exports: {} } })
   const script = new vm.Script(source)
   script.runInContext(context)
   return context.module.exports
}


types.ts
export type Formatter = (input: string | number | null | undefined) => string

export interface Colors {
   isColorSupported: boolean
   reset: Formatter
   bold: Formatter
   dim: Formatter
   italic: Formatter
   underline: Formatter
   inverse: Formatter
   hidden: Formatter
   strikethrough: Formatter
   black: Formatter
   red: Formatter
   green: Formatter
   yellow: Formatter
   blue: Formatter
   magenta: Formatter
   cyan: Formatter
   white: Formatter
   gray: Formatter
   bgBlack: Formatter
   bgRed: Formatter
   bgGreen: Formatter
   bgYellow: Formatter
   bgBlue: Formatter
   bgMagenta: Formatter
   bgCyan: Formatter
   bgWhite: Formatter
   nyxbpurple: Formatter
   nyxbgreen: Formatter
   nyxbcyan: Formatter
   nyxbblue: Formatter
   nyxbyellow: Formatter
   nyxbfox: Formatter
   nyxbred: Formatter
}


benchmarks/size.js
#!/usr/bin/env node

const { get } = require('node:https')

const { bold, gray } = require('../colors.js')

async function getJSON(url) {
   return new Promise((resolve) => {
      get(url, (res) => {
         let text = ''
         res.on('data', (chunk) => {
            text += chunk
         })
         res.on('end', () => {
            resolve(JSON.parse(text))
         })
      })
   })
}

async function benchmark(lib) {
   const data = await getJSON(`https://packagephobia.com/v2/api.json?p=${lib}`)
   const size = data.install.bytes
   process.stdout.write(
      `${lib.padEnd('ansi-colors  '.length)
         + bold(
            Math.round(size / 1024)
               .toString()
               .padStart(4),
         )
         } kB\n`,
   )
}

async function start() {
   process.stdout.write(gray('Data from packagephobia.com\n'))
   await benchmark('  chalk')
   await benchmark('  cli-color')
   await benchmark('  ansi-colors')
   await benchmark('  kleur')
   await benchmark('  colorette')
   await benchmark('  nanocolors')
   await benchmark('+ colors')
}

start()


benchmarks/simple.js
#!/usr/bin/env node

// Benchmark results are unstable. To have more stable results:
// 1. Restart OS. Do not run any applications. Put power cable to laptop.
// 2. Run tests 5 times.
// 3. Took the best result for each candidate.

const benchmark = require('benchmark')
const colorette = require('colorette')
const kleur = require('kleur')
const kleurColors = require('kleur/colors')
const chalk = require('chalk')
const ansi = require('ansi-colors')
const cliColor = require('cli-color')
const nanocolors = require('nanocolors')
const colors = require('../colors.js')

function formatNumber(number) {
   return String(number)
      .replace(/\d{3}$/, ',$&')
      .replace(/^(\d|\d\d)(\d{3},)/, '$1,$2')
}

console.log(colorette.green('colorette'))
console.log(kleur.green('kleur'))
console.log(chalk.green('chalk'))
console.log(ansi.green('ansi'))
console.log(cliColor.green('cliColor'))
console.log(colors.green('colors'))
console.log(nanocolors.green('nanocolors'))

const suite = new benchmark.Suite()
let out

suite
   .add('chalk', () => {
      out = chalk.red('Add plugin to use time limit')
   })
   .add('cli-color', () => {
      out = cliColor.red('Add plugin to use time limit')
   })
   .add('ansi-colors', () => {
      out = ansi.red('Add plugin to use time limit')
   })
   .add('kleur', () => {
      out = kleur.red('Add plugin to use time limit')
   })
   .add('kleur/colors', () => {
      out = kleurColors.red('Add plugin to use time limit')
   })
   .add('colorette', () => {
      out = colorette.red('Add plugin to use time limit')
   })
   .add('nanocolors', () => {
      out = nanocolors.red('Add plugin to use time limit')
   })
   .add('colors', () => {
      out = colors.red('Add plugin to use time limit')
   })
   .on('cycle', (event) => {
      const prefix = event.target.name === 'colors' ? '+ ' : '  '
      const name = event.target.name.padEnd('kleur/colors  '.length)
      const hz = formatNumber(event.target.hz.toFixed(0)).padStart(10)
      process.stdout.write(`${prefix}${name}${colors.bold(hz)} ops/sec\n`)
   })
   .on('error', (event) => {
      process.stderr.write(`${colors.red(event.target.error.toString())}\n`)
      process.exit(1)
   })
   .run()


benchmarks/complex.js
#!/usr/bin/env node

// Benchmark results are unstable. To have more stable results:
// 1. Restart OS. Do not run any applications. Put power cable to laptop.
// 2. Run tests 5 times.
// 3. Took the best result for each candidate.

const benchmark = require('benchmark')
const colorette = require('colorette')
const kleur = require('kleur')
const kleurColors = require('kleur/colors')
const chalk = require('chalk')
const ansi = require('ansi-colors')
const cliColor = require('cli-color')
const nanocolors = require('nanocolors')
const colors = require('../colors.js')

function formatNumber(number) {
   return String(number)
      .replace(/\d{3}$/, ',$&')
      .replace(/^(\d|\d\d)(\d{3},)/, '$1,$2')
}

const suite = new benchmark.Suite()
let out

let index = 1e8

suite
   .add('chalk', () => {
      out
         = chalk.red('.')
         + chalk.yellow('.')
         + chalk.green('.')
         + chalk.bgRed(chalk.black(' ERROR '))
         + chalk.red(
            ` Add plugin ${chalk.yellow('name')} to use time limit with ${chalk.yellow(++index)}`,
         )
   })
   .add('cli-color', () => {
      out
         = cliColor.red('.')
         + cliColor.yellow('.')
         + cliColor.green('.')
         + cliColor.bgRed(cliColor.black(' ERROR '))
         + cliColor.red(
            ` Add plugin ${
                cliColor.yellow('name')
                } to use time limit with ${
                cliColor.yellow(++index)}`,
         )
   })
   .add('ansi-colors', () => {
      out
         = ansi.red('.')
         + ansi.yellow('.')
         + ansi.green('.')
         + ansi.bgRed(ansi.black(' ERROR '))
         + ansi.red(
            ` Add plugin ${ansi.yellow('name')} to use time limit with ${ansi.yellow(++index)}`,
         )
   })
   .add('kleur', () => {
      out
         = kleur.red('.')
         + kleur.yellow('.')
         + kleur.green('.')
         + kleur.bgRed(kleur.black(' ERROR '))
         + kleur.red(
            ` Add plugin ${kleur.yellow('name')} to use time limit with ${kleur.yellow(++index)}`,
         )
   })
   .add('kleur/colors', () => {
      out
         = kleurColors.red('.')
         + kleurColors.yellow('.')
         + kleurColors.green('.')
         + kleurColors.bgRed(kleurColors.black(' ERROR '))
         + kleurColors.red(
            ` Add plugin ${
                kleurColors.yellow('name')
                } to use time limit with ${
                kleurColors.yellow(++index)}`,
         )
   })
   .add('colorette', () => {
      out
         = colorette.red('.')
         + colorette.yellow('.')
         + colorette.green('.')
         + colorette.bgRed(colorette.black(' ERROR '))
         + colorette.red(
            ` Add plugin ${
                colorette.yellow('name')
                } to use time limit with ${
                colorette.yellow(++index)}`,
         )
   })
   .add('nanocolors', () => {
      out
         = nanocolors.red('.')
         + nanocolors.yellow('.')
         + nanocolors.green('.')
         + nanocolors.bgRed(nanocolors.black(' ERROR '))
         + nanocolors.red(
            ` Add plugin ${
                nanocolors.yellow('name')
                } to use time limit with ${
                nanocolors.yellow(++index)}`,
         )
   })
   .add('colors', () => {
      out
      = colors.red('.')
      + colors.yellow('.')
      + colors.green('.')
      + colors.bgRed(colors.black(' ERROR '))
      + colors.red(
         ` Add plugin ${
             colors.yellow('name')
                } to use time limit with ${
                colors.yellow(`${++index}`)}`,
      )
   })
   .on('cycle', (event) => {
      const prefix = event.target.name === 'colors' ? '+ ' : '  '
      const name = event.target.name.padEnd('kleur/colors  '.length)
      const hz = formatNumber(event.target.hz.toFixed(0)).padStart(10)
      process.stdout.write(`${prefix}${name}${colors.bold(hz)} ops/sec\n`)
   })
   .on('error', (event) => {
      process.stderr.write(`${colors.red(event.target.error.toString())}\n`)
      process.exit(1)
   })
   .run()


benchmarks/loading.js
#!/usr/bin/env node

const { execSync } = require('node:child_process')

const RUNS = 50

const results = {}

for (let i = 0; i < RUNS; i++) {
   const output = execSync('node ./benchmarks/loading-runner.js').toString()
   output
      .trim()
      .split('\n')
      .forEach((line) => {
         const [name, result] = line.split(' ')
         results[name] = (results[name] || 0) + parseFloat(result)
      })
}

for (const name in results) {
   const prefix = name === 'colors' ? '+ ' : '  '
   const title = name.padEnd('kleur/colors  '.length)
   const time = (Math.round((1000 * results[name]) / RUNS) / 1000)
      .toString()
      .replace(/\.\d$/, '$&00')
      .replace(/\.\d\d$/, '$&0')
   process.stdout.write(`${prefix + title}\x1B[1m${time.padStart(6)}\x1B[22m ms\n`)
}


benchmarks/loading-runner.js
const { performance } = require('node:perf_hooks')

let before
function showTime(name) {
   const after = performance.now()
   process.stdout.write(`${name} ${after - before}\n`)
}

before = performance.now()
const chalk = require('chalk')

showTime('chalk')

before = performance.now()
const cliColor = require('cli-color')

showTime('cli-color')

before = performance.now()
const ansi = require('ansi-colors')

showTime('ansi-colors')

before = performance.now()
const kleur = require('kleur')

showTime('kleur')

before = performance.now()
const kleurColors = require('kleur/colors')

showTime('kleur/colors')

before = performance.now()
const colorette = require('colorette')

showTime('colorette')

before = performance.now()
const nanocolors = require('nanocolors')

showTime('nanocolors')

before = performance.now()
const colors = require('../colors.js')

showTime('colors')


colors.js
const tty = require('node:tty')

const isColorSupported
   = !('NO_COLOR' in process.env || process.argv.includes('--no-color'))
   && ('FORCE_COLOR' in process.env
      || process.argv.includes('--color')
      || process.platform === 'win32'
      || (tty.isatty(1) && process.env.TERM !== 'dumb')
      || 'CI' in process.env)

function formatter(open, close, replace = open) {
   return (input) => {
      const string = `${input}`
      const index = string.indexOf(close, open.length)
      return ~index
         ? open + replaceClose(string, close, replace, index) + close
         : open + string + close
   }
}

let replaceClose = (string, close, replace, index) => {
   const start = string.substring(0, index) + replace
   const end = string.substring(index + close.length)
   const nextIndex = end.indexOf(close)
   return ~nextIndex ? start + replaceClose(end, close, replace, nextIndex) : start + end
}

function createColors(enabled = isColorSupported) {
   return {
      isColorSupported: enabled,
      reset: enabled ? s => `\x1B[0m${s}\x1B[0m` : String,
      bold: enabled ? formatter('\x1B[1m', '\x1B[22m', '\x1B[22m\x1B[1m') : String,
      dim: enabled ? formatter('\x1B[2m', '\x1B[22m', '\x1B[22m\x1B[2m') : String,
      italic: enabled ? formatter('\x1B[3m', '\x1B[23m') : String,
      underline: enabled ? formatter('\x1B[4m', '\x1B[24m') : String,
      inverse: enabled ? formatter('\x1B[7m', '\x1B[27m') : String,
      hidden: enabled ? formatter('\x1B[8m', '\x1B[28m') : String,
      strikethrough: enabled ? formatter('\x1B[9m', '\x1B[29m') : String,
      black: enabled ? formatter('\x1B[30m', '\x1B[39m') : String,
      red: enabled ? formatter('\x1B[31m', '\x1B[39m') : String,
      green: enabled ? formatter('\x1B[32m', '\x1B[39m') : String,
      yellow: enabled ? formatter('\x1B[33m', '\x1B[39m') : String,
      blue: enabled ? formatter('\x1B[34m', '\x1B[39m') : String,
      magenta: enabled ? formatter('\x1B[35m', '\x1B[39m') : String,
      cyan: enabled ? formatter('\x1B[36m', '\x1B[39m') : String,
      white: enabled ? formatter('\x1B[37m', '\x1B[39m') : String,
      gray: enabled ? formatter('\x1B[90m', '\x1B[39m') : String,
      bgBlack: enabled ? formatter('\x1B[40m', '\x1B[49m') : String,
      bgRed: enabled ? formatter('\x1B[41m', '\x1B[49m') : String,
      bgGreen: enabled ? formatter('\x1B[42m', '\x1B[49m') : String,
      bgYellow: enabled ? formatter('\x1B[43m', '\x1B[49m') : String,
      bgBlue: enabled ? formatter('\x1B[44m', '\x1B[49m') : String,
      bgMagenta: enabled ? formatter('\x1B[45m', '\x1B[49m') : String,
      bgCyan: enabled ? formatter('\x1B[46m', '\x1B[49m') : String,
      bgWhite: enabled ? formatter('\x1B[47m', '\x1B[49m') : String,
      nyxbpurple: enabled ? formatter('\x1B[38;2;153;69;255m', '\x1B[39m') : String,
      nyxbgreen: enabled ? formatter('\x1B[38;2;20;241;149m', '\x1B[39m') : String,
      nyxbcyan: enabled ? formatter('\x1B[38;2;0;255;255m', '\x1B[39m') : String,
      nyxbblue: enabled ? formatter('\x1B[38;2;88;101;242m', '\x1B[39m') : String,
      nyxbyellow: enabled ? formatter('\x1B[38;2;254;231;92m', '\x1B[39m') : String,
      nyxbfox: enabled ? formatter('\x1B[38;2;235;69;158m', '\x1B[39m') : String,
      nyxbred: enabled ? formatter('\x1B[38;2;237;66;69m', '\x1B[39m') : String,
   }
}

module.exports = createColors()
module.exports.createColors = createColors
